src/langchain_google_spanner/graph_store.py

```diff
@@ -0,0 +1,1176 @@
+from langchain_community.graphs.graph_store import GraphStore
+from requests.structures import CaseInsensitiveDict
+
+
+def to_identifier(s: str) -> str:
+    return "`" + s + "`"
+
+
+def to_identifiers(s: List[str]) -> str:
+    return map(to_identifier, s)
+
+
+class NodeWrapper(object):
+    """Wrapper around Node to support set operations using node id"""
+
+    def __init__(self, node: Node):
+        self.node = node
+
+    def __hash__(self):
+        return hash(self.node.id)
+
+    def __eq__(self, other: Any):
+        if isinstance(other, NodeWrapper):
+            return self.node.id == other.node.id
+        return False
+
+
+class EdgeWrapper(object):
+    """Wrapper around Relationship to support set operations using edge source and target id"""
+
+    def __init__(self, edge: Relationship):
+        self.edge = edge
+
+    def __hash__(self):
+        return hash(
+            (
+                self.edge.source.id,
+                self.edge.target.id,
+            )
+        )
+
+    def __eq__(self, other: Any):
+        if isinstance(other, EdgeWrapper):
+            return (
+                self.edge.source.id == other.edge.source.id
+                and self.edge.target.id == other.edge.target.id
+            )
+        return False
+
+
+def partition_graph_docs(
+    graph_documents: List[GraphDocument],
+) -> Tuple(dict, dict):
+    """Returns nodes and edges grouped by the type.
+
+    Parameters:
+    - graph_documents: List[GraphDocument]
+
+    Returns:
+    - A tuple of two dictionaries. The first is nodes grouped by node types,
+      the second is edges grouped by edge types.
+    """
+    nodes = CaseInsensitiveDict()
+    edges = CaseInsensitiveDict()
+    for doc in graph_documents:
+        for node in doc.nodes:
+            s = nodes.setdefault(node.type, set())
+            w = NodeWrapper(node)
+            if w in s:
+                # Combine the properties for nodes with the same id.
+                n = next(filter(lambda v: v == w, s))
+                n.node.properties.update(node.properties)
+            else:
+                s.add(w)
+
+        for edge in doc.relationships:
+            # Partition edges by the triplet because there could be edges with the
+            # same type but between different types of nodes.
+            edge_name = "{}_{}_{}".format(edge.source.type, edge.type, edge.target.type)
+            s = edges.setdefault(edge_name, set())
+            w = EdgeWrapper(edge)
+            if w in s:
+                # Combine the properties for edges with the same id.
+                e = next(filter(lambda v: v == w, s))
+                e.edge.properties.update(edge.properties)
+            else:
+                s.add(w)
+    return {name: [w.node for w in ws] for name, ws in nodes.items()}, {
+        name: [w.edge for w in ws] for name, ws in edges.items()
+    }
+
+
+class TypeUtility(object):
```

Comment:
This is very generic class and not limited just to GraphStore.  Can we create a utility module and move it there? It would be useful for future development in this repo.

Reply:
Done!